### ISSUE TITLE
Geometry field cannot be converted to or from Oracle data type

### ISSUE THREAD

**Issue by fnicollet** [laurenz/oracle_fdw#91]
Hello,

I am running latest PG 9.5 (9.5.3), Postgis 2.2 , oracle_fdw-1.4.0-pg95-win64 that I downloaded from Github, with Oracle instantclient-basic-windows.x64-12.1.0.2.0. This is all running on my machine on Windows 10 64-Bits.
I used IMPORT FOREIGN SCHEMA and it referenced one of my Oracle table which contains a geometry in a field called "geometrie". When I run a simple SELECT on the FDT, I get the following:
![image](https://cloud.githubusercontent.com/assets/1240481/15427845/0c070760-1e98-11e6-9f7e-1066a6d98fb3.png)

The column is of type "text" and is empty, while my Oracle table has LINESTRING geometries.
My Oracle table is correctly referenced in USER_SRO_METADATA, the field is of type SDO_GEOMETRY:
![image](https://cloud.githubusercontent.com/assets/1240481/15427864/3b95244e-1e98-11e6-8e55-0e7940d13991.png)

If instead of IMPORT FOREIGN SCHEMA, I manually create a foreign table, with the field as:
>> geometrie geometry

If I run
>> select geometrie from test_itin where code_itin = 2;

I get an error from the FDW:
```
ERREUR:  column "geometrie" of foreign table "test_itin" cannot be converted to or from Oracle data type
```
Note that I also tried geometrie geometry(LINESTRING, 2154) which gives the same result

I had look at the error and found this bit of the source code:
https://github.com/laurenz/oracle_fdw/blob/817c44818b319bfb038739de3cbba35ca634ac0f/oracle_fdw.c#L3916
It seems like this test:
>> if (oratype == ORA_TYPE_GEOMETRY && pgtype == GEOMETRYOID)

Fails for some reason in my case.

I'm a bit stuck here, do you have any idea what's wrong with what I am doing?

Thanks,
Fabien

**Comment by laurenz** [comment]
It could either be that oracle_fdw does not recognize the Oracle data type (that would be a bug in oracle_fdw) or that it does not recognize your postgis installation.

The latter can easily be tested in PostgreSQL.  
Start a new session and enter:

    test=> SET client_min_messages=debug1;
    test=> SELECT oracle_diag();
    DEBUG:  oracle_fdw: PostGIS is installed, GEOMETRYOID = 16399
                                            oracle_diag
    ----------------------------------------------------------------------------------------------
     oracle_fdw 1.4.0, PostgreSQL 9.6devel, Oracle client 11.2.0.3.0, ORACLE_HOME=/magwien/oracle
    (1 row)

See if you get a debug message like the above!

**Comment by fnicollet** [comment]
Here is what I get with these 2 queries:
temp=# SET client_min_messages=debug1;
SET
temp=# SELECT oracle_diag();
DEBUG:  oracle_fdw: PostGIS is installed, GEOMETRYOID = 33766
                         oracle_diag
--------------------------------------------------------------
 oracle_fdw 1.4.0, PostgreSQL 9.5.3, Oracle client 12.1.0.2.0
(1 ligne)

**Comment by laurenz** [comment]
Can you start sqlplus and run

    DESC <tablename>

and tell me what it shows for the geometry column?

If that doesn't clear up things, I'll send you a patch that will write out some debug info.

**Comment by fnicollet** [comment]
I get "MDSYS.SDO_GEOMETRY" for the geometry column (the whole describe is in the other issue I posted today: https://github.com/laurenz/oracle_fdw/issues/90)

**Comment by laurenz** [comment]
Oh, then the explanation is simple.

The column `geometrie` in your foreign table is of type `text`, and there is no mapping between these two types.
Define `geometrie` as type `geometry`; you can use `ALTER FOREIGN TABLE` to change the table definition.

One possible explanation would be that you created the foreign table with IMPORT FOREIGN SCHEMA, but PostGIS was not installed at the time.

**Comment by fnicollet** [comment]
Wow, that was it, I created and initialized the oracle_fdw extension before create extension postgis (I was so eager to give oracle_fdw a try !)
I started the db from scratch , created the postgis extension, then the oracle_fdw extension and got USER-DEFINED instead of text. And now, I can select the geometries without any problem

Again, thanks a lot for your help!
Fabien

**Comment by GregStef** [comment]
Hi laurenz
First of all, *a very big THANK YOU* for your contribution to the community!

I've been working with Oracle Spatial for about 15 years and this is my first attempt to mess around with PG.

I have exactly the same problem as fnikollet had, but all settings are as they're supposed to be (I think):
`SELECT version();`
EnterpriseDB 9.5.3.8 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.1.2 20080704 (Red Hat 4.1.2-55), 64-bit

`SELECT PostGIS_Full_Version();`
POSTGIS="2.1.8 r13780" GEOS="3.4.2-CAPI-1.8.2 r3921" PROJ="Rel. 4.9.1, 04 March 2015" GDAL="GDAL 1.11.2, released 2015/02/10 GDAL_DATA not found" LIBXML="2.9.2" LIBJSON="UNKNOWN" RASTER

PostGIS was installed about a month ago - oracle_fdw was installed yesterday.

`SET client_min_messages=debug1;`
`select sysdate,oracle_diag;`
2016-06-17 10:53:21   oracle_fdw 1.4.0, PostgreSQL 9.5.3.8, Oracle client 11.2.0.4.0

After creating oracle_fdw extension (done by our DBA), I ran:
`CREATE SERVER gaeedb_ora FOREIGN DATA WRAPPER oracle_fdw  OPTIONS (dbserver '//<the_IP>/GAEEDB');`

`GRANT USAGE ON FOREIGN SERVER gaeedb_ora TO greg;`

`CREATE USER MAPPING FOR greg SERVER gaeedb_ora  OPTIONS (user '<the_ora_user>', password '<the_pass>');`

My test_query is: `select * from gregstef_r.gr_pe15 where geom is not null;`

Attempt 1:
`import foreign schema "GREGSTEF" from server gaeedb_ora into gregstef_r;`
It successfully imported all definitions, BUT all geometry columns in all foreign tables were of TEXT datatype.
The test_query returns all rows, although the column 'geom' is empty.

Attempt 2:
`ALTER FOREIGN TABLE gregstef_r.gr_pe15 ALTER COLUMN geom SET DATA type public.geometry;`
The test_query returns:
ERROR: column "geom" of foreign table "gr_pe15" cannot be converted to or from Oracle data type

Attempt 3:
`CREATE FOREIGN TABLE gregstef_r.gr_pe15 (
    id numeric OPTIONS (key 'true')  NOT NULL,
    geom public.geometry
)
SERVER gaeedb_ora OPTIONS (readonly 'true', schema 'GREGSTEF', table 'GR_PE15');`
The test_query returns the same error.

I also tried a variation of Attempt3, setting geom as public.geometry(polygon,2100) - same result.
I also tried with `SERVER gaeedb_ora OPTIONS (readonly 'true', table '(SELECT id, geom FROM GREGSTEF.GR_PE15 where rownum < 11)') ;`

Some additional info:
1) User "greg" *is* a superuser, but *not* the user who owns the postgres installation.
2) The foreign server was set with an IP, not a hostname
3) All attempts made with oracle tables containing either only simple polygons, or a mix of simple and multi polygons - no lines, points, compounds on anything else weird.
4) Within the PG database I can successfully create and populate other tables with the same data, from the same oracle (using FME, QGIS etc).

ANY help would be immensely appreciated...

Thanks in advance for your time
Greg

**Comment by fnicollet** [comment]
Before @laurenz answers, when you run
```
temp=# SELECT oracle_diag();
```
Do you not get  this?
```
DEBUG:  oracle_fdw: PostGIS is installed, GEOMETRYOID = 33766
                         oracle_diag
```
When you say that postgis was installed, do you mean that it was installed as part of the postgresql install or did you also run "create extension postgis;"  on the database a month ago?

**Comment by GregStef** [comment]
Hi Fabien

I run

SET client_min_messages=debug1;
SELECT oracle_diag();

from within PGAdminIII.
I get the same if I run from DBeaver (the tool I usually use).
Do I have to run it in a terminal?  I'm a newbie with PG...

Postgis is an old thing already - I've being using it for more than a month
already, so, to answer your question, the extension has been created since
over a month.

**Comment by GregStef** [comment]
Ah, sorry, I replied in gmail. 
To keep everything in one place, here it is:
Hi Fabien

I run

SET client_min_messages=debug1;
SELECT oracle_diag();

from within PGAdminIII.
I get the same if I run from DBeaver (the tool I usually use).
Do I have to run it in a terminal?  I'm a newbie with PG...

Postgis is an old thing already - I've being using it for more than a month already, so, to answer your question, the extension has been created since over a month.

**Comment by GregStef** [comment]
And running from a terminal:
[enterprisedb@gisdb ~]$ psql -U greg -d gregdb151
psql.bin (9.5.3.8)
Type "help" for help.

gregdb151=# SET client_min_messages=debug1;
SET
gregdb151=# SELECT oracle_diag();
                          oracle_diag
----------------------------------------------------------------
 oracle_fdw 1.4.0, PostgreSQL 9.5.3.8, Oracle client 11.2.0.4.0
(1 row)

gregdb151=#

**Comment by fnicollet** [comment]
ah yes, the SET commands don't work in pgAdmin, probably because it is an
environment variable or something similar. I had to try it using psql,
which is installed by default on the server

On Fri, Jun 17, 2016 at 2:16 PM, GregStef <notifications@github.com> wrote:

> Hi Fabien
>
> I run
>
> SET client_min_messages=debug1;
> SELECT oracle_diag();
>
> from within PGAdminIII.
> I get the same if I run from DBeaver (the tool I usually use).
> Do I have to run it in a terminal? I'm a newbie with PG...
>
> Postgis is an old thing already - I've being using it for more than a month
> already, so, to answer your question, the extension has been created since
> over a month.
>
> On Fri, Jun 17, 2016 at 3:00 PM, Fabien NICOLLET <notifications@github.com
> >
> wrote:
>
> > Before @laurenz <https://github.com/laurenz> answers, when you run
> >
> > temp=# SELECT oracle_diag();
> >
> > Do you not get this?
> >
> > DEBUG: oracle_fdw: PostGIS is installed, GEOMETRYOID = 33766
> > oracle_diag
> >
> > When you say that postgis was installed, do you mean that it was
> installed
> > as part of the postgresql install or did you also run "create extension
> > postgis;" on the database a month ago?
> >
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/laurenz/oracle_fdw/issues/91#issuecomment-226751511
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe/ATEUSe-ohKGgcp3lVcf5dln30pElDHPhks5qMoxUgaJpZM4IjKgZ
> >
> > .
>
> >
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/laurenz/oracle_fdw/issues/91#issuecomment-226754139>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABLtobUiBzydSz7i7L1fJwpJFiiU1raNks5qMpAfgaJpZM4IjKgZ>
> .
>



-- 
Fabien Nicollet
Tel : +33 (0)6 60 54 72 20

**Comment by GregStef** [comment]
As posted, I get the same response from the terminal.

Also, I think There is an item (5) to be added to the additional info (original post): 
5) I'm working in an EnterpriseDB server

**Comment by fnicollet** [comment]
ok, then i will let @laurenz answer, I think there should be the postgis log message in there, not sure why there isn't but it's probably the source of your issues

**Comment by GregStef** [comment]
Thanks Fabien

**Comment by laurenz** [comment]
Thanks @fnicollet, you are spot on.

The `SET`command works just fine in pgAdmin III.
After running `oracle_diag`, you will have to click the 'Messages' tab in the 'Output Pane' to see the debug messages.
But if the message does not show up with `psql` either, it seems like oracle_fdw cannot find the PostGIS installation.

What do you get when you run the following in `psql`:

    SELECT n.nspname, f.prorettype::regtype, f.proargtypes
    FROM pg_namespace n
         JOIN pg_proc f ON n.oid = f.pronamespace
    WHERE proname = 'geometry_recv';

On a working system you should get the following:

     nspname | prorettype | proargtypes
    ---------+------------+-------------
     public  | geometry   | 2281
    (1 row)

If there is no result or more than one result, that would explain your problem.

There used to be a bug where PostGIS could not be detected when ArcGIS was installed, but that is fixed in 1.4.0.

**Comment by GregStef** [comment]
Hmmm.. this is not going to be easy...

The psql output:
` nspname | prorettype | proargtypes
---------+------------+-------------
 public  | geometry   | 2281
(1 row)
`

**Comment by GregStef** [comment]
nspname | prorettype | proargtypes
---------+------------+-------------
 public  | geometry   | 2281
(1 row)

**Comment by laurenz** [comment]
Could you apply the diagnostic patch in [this ZIP file](https://github.com/laurenz/oracle_fdw/files/320539/postgis.zip), build oracle_fdw, start a new psql session and run:

    test=> SELECT oracle_diag();

On a working system, you'll get:

    DEBUG:  oracle_fdw: entering _PG_init
    DEBUG:  oracle_fdw: Found 1 copies of "geometry_recv(internal)"
    DEBUG:  oracle_fdw: PostGIS is installed, GEOMETRYOID = 16463
                              oracle_diag
    ----------------------------------------------------------------
     oracle_fdw 1.4.0, ...
    (1 row)

I'd like to know what you see on your system.

**Comment by GregStef** [comment]
Thanks @laurenz - I'll hand the zip over to our DBA/SO and get back at you

**Comment by GregStef** [comment]
Hi @laurenz 
Here are the results:
[enterprisedb@gisdb ~]$ psql -U greg -d gregdb151
psql.bin (9.5.3.8)
Type "help" for help.

gregdb151=# SET client_min_messages=debug1;
SET
gregdb151=# SELECT oracle_diag();
DEBUG:  oracle_fdw: entering _PG_init
DEBUG:  oracle_fdw: Found 0 copies of "geometry_recv(internal)"
                          oracle_diag
----------------------------------------------------------------
 oracle_fdw 1.4.0, PostgreSQL 9.5.3.8, Oracle client 11.2.0.4.0
(1 row)

gregdb151=#

What do you make of this?

**Comment by GregStef** [comment]
Laurenz, the colleague who set the whole thing up, told me that he got a few warnings during build.
He can't tell if they are significant, but here they are:
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -g -DLINUX_OOM_ADJ=0 -O2 -DMAP_HUGETLB=0x40000 -fpic -I/sdk/include -I/oci/include -I/rdbms/public -I/usr/include/oracle/12.1/client -I/usr/include/oracle/12.1/client64 -I/usr/include/oracle/11.2/client -I/usr/include/oracle/11.2/client64 -I/usr/include/oracle/11.1/client -I/usr/include/oracle/11.1/client64 -I/usr/include/oracle/10.2.0.5/client -I/usr/include/oracle/10.2.0.5/client64 -I/usr/include/oracle/10.2.0.4/client -I/usr/include/oracle/10.2.0.4/client64 -I/usr/include/oracle/10.2.0.3/client -I/usr/include/oracle/10.2.0.3/client64 -I. -I./ -I/opt/PostgresPlus/9.5AS/include/server -I/opt/PostgresPlus/9.5AS/include/internal -D_GNU_SOURCE -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include/libxml2 -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include   -I/opt/local/Current/include/libxml2 -I/opt/local/Current/include  -c -o oracle_fdw.o oracle_fdw.c
oracle_fdw.c: In function ‘oracleRegisterCallback’:
oracle_fdw.c:5170: warning: passing argument 1 of ‘RegisterXactCallback’ from incompatible pointer type
/opt/PostgresPlus/9.5AS/include/server/access/xact.h:354: note: expected ‘XactCallback’ but argument is of type ‘void (*)(enum XactEvent,  void *)’
oracle_fdw.c:5172: warning: passing argument 1 of ‘RegisterSubXactCallback’ from incompatible pointer type
/opt/PostgresPlus/9.5AS/include/server/access/xact.h:358: note: expected ‘SubXactCallback’ but argument is of type ‘void (*)(enum SubXactEvent,  SubTransactionId,  SubTransactionId,  void *)’
oracle_fdw.c: In function ‘oracleUnregisterCallback’:
oracle_fdw.c:5183: warning: passing argument 1 of ‘UnregisterXactCallback’ from incompatible pointer type
/opt/PostgresPlus/9.5AS/include/server/access/xact.h:356: note: expected ‘XactCallback’ but argument is of type ‘void (*)(enum XactEvent,  void *)’
oracle_fdw.c:5185: warning: passing argument 1 of ‘UnregisterSubXactCallback’ from incompatible pointer type
/opt/PostgresPlus/9.5AS/include/server/access/xact.h:359: note: expected ‘SubXactCallback’ but argument is of type ‘void (*)(enum SubXactEvent,  SubTransactionId,  SubTransactionId,  void *)’
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -g -DLINUX_OOM_ADJ=0 -O2 -DMAP_HUGETLB=0x40000 -fpic -I/sdk/include -I/oci/include -I/rdbms/public -I/usr/include/oracle/12.1/client -I/usr/include/oracle/12.1/client64 -I/usr/include/oracle/11.2/client -I/usr/include/oracle/11.2/client64 -I/usr/include/oracle/11.1/client -I/usr/include/oracle/11.1/client64 -I/usr/include/oracle/10.2.0.5/client -I/usr/include/oracle/10.2.0.5/client64 -I/usr/include/oracle/10.2.0.4/client -I/usr/include/oracle/10.2.0.4/client64 -I/usr/include/oracle/10.2.0.3/client -I/usr/include/oracle/10.2.0.3/client64 -I. -I./ -I/opt/PostgresPlus/9.5AS/include/server -I/opt/PostgresPlus/9.5AS/include/internal -D_GNU_SOURCE -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include/libxml2 -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include   -I/opt/local/Current/include/libxml2 -I/opt/local/Current/include  -c -o oracle_utils.o oracle_utils.c
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -g -DLINUX_OOM_ADJ=0 -O2 -DMAP_HUGETLB=0x40000 -fpic -I/sdk/include -I/oci/include -I/rdbms/public -I/usr/include/oracle/12.1/client -I/usr/include/oracle/12.1/client64 -I/usr/include/oracle/11.2/client -I/usr/include/oracle/11.2/client64 -I/usr/include/oracle/11.1/client -I/usr/include/oracle/11.1/client64 -I/usr/include/oracle/10.2.0.5/client -I/usr/include/oracle/10.2.0.5/client64 -I/usr/include/oracle/10.2.0.4/client -I/usr/include/oracle/10.2.0.4/client64 -I/usr/include/oracle/10.2.0.3/client -I/usr/include/oracle/10.2.0.3/client64 -I. -I./ -I/opt/PostgresPlus/9.5AS/include/server -I/opt/PostgresPlus/9.5AS/include/internal -D_GNU_SOURCE -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include/libxml2 -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include -I/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/include   -I/opt/local/Current/include/libxml2 -I/opt/local/Current/include  -c -o oracle_gis.o oracle_gis.c
gcc -Wall -Wmissing-prototypes -Wpointer-arith -Wdeclaration-after-statement -Wendif-labels -Wmissing-format-attribute -Wformat-security -fno-strict-aliasing -fwrapv -g -DLINUX_OOM_ADJ=0 -O2 -DMAP_HUGETLB=0x40000 -fpic -shared -o oracle_fdw.so oracle_fdw.o oracle_utils.o oracle_gis.o -L/opt/PostgresPlus/9.5AS/lib -L/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/lib -L/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/lib  -L/opt/local/Current/lib -L/opt/local/20150616/181474c6-228d-4d58-b663-17981352ebce/lib  -Wl,--as-needed -Wl,-rpath,'/opt/PostgresPlus/9.5AS/lib',--enable-new-dtags  -L -L/bin -L/lib -lclntsh -L/usr/lib/oracle/12.1/client/lib -L/usr/lib/oracle/12.1/client64/lib -L/usr/lib/oracle/11.2/client/lib -L/usr/lib/oracle/11.2/client64/lib -L/usr/lib/oracle/11.1/client/lib -L/usr/lib/oracle/11.1/client64/lib -L/usr/lib/oracle/10.2.0.5/client/lib -L/usr/lib/oracle/10.2.0.5/client64/lib -L/usr/lib/oracle/10.2.0.4/client/lib -L/usr/lib/oracle/10.2.0.4/client64/lib -L/usr/lib/oracle/10.2.0.3/client/lib -L/usr/lib/oracle/10.2.0.3/client64/lib

**Comment by laurenz** [comment]
These warnings are probably because you are using EDB's PostgreSQL fork, which I am not familiar with.
They sound disturbing.
How is `XactCallback` and `SubXactCallback` defined in `xact.h`?

I don't get how an index scan in `_PG_init` cannort find the record when an SQL query can.

As a last-ditch effort, on the suspicion that you have a corrupt index on a system table, try the following as superuser:

    REINDEX INDEX pg_catalog.pg_proc_proname_args_nsp_index;

and see if that helps.

If that does not help, I'd suspect that something is subtly different in EDB's PostgreSQL fork &hellip;
Is there a chance you could ask EDB for help?
Maybe they can tell why the code [here](https://github.com/laurenz/oracle_fdw/blob/master/oracle_fdw.c#L628) returns zero results when an SQL query works.

**Comment by GregStef** [comment]
Regarding  `xact.h` , we'll have to wait for our DBA to get an answer.
Regarding reindexing, after spawning a new connection, I get same results:
`gregdb151=# SET client_min_messages=debug1;`
SET
`gregdb151=# SELECT oracle_diag();`
DEBUG:  oracle_fdw: entering _PG_init
DEBUG:  oracle_fdw: Found 0 copies of "geometry_recv(internal)"
                          oracle_diag
----------------------------------------------------------------
 oracle_fdw 1.4.0, PostgreSQL 9.5.3.8, Oracle client 11.2.0.4.0
(1 row)

Our DBA has gone for the weekend, so I cannot answer your question - we'll have to wait until Tuesday (Monday is a public holiday here).

If you have any other suggestions/ideas, I'm more than willing to try!

**Comment by laurenz** [comment]
I'm afraid that I am at my wit's end here; I strongly suspect that has to do with differences in EDB's PostgreSQL fork.

**Comment by GregStef** [comment]
I'm inclined to agree here - there seems to be no way out. 

My workaround, for what's worth, is: Create one oracle view per table, in which, instead of selecting the geometry column, I put its WKB equivalent. Then, in PG, I reconstitute the geometry (with its SRID, since WKB doesn't have one) and I proceed from there. Messy and costly, for sure, but it's the only thing I could come up with thus far.

Everything else I've tried has failed, whereas  using oracle_fdw with the same source data but in a non-EDB installation, presents no problem whatsoever.
Pity!

Anyway, thank you very much for your time and effort.
Best regards
Greg
